lib/spack/spack/spec.py

```diff
@@ -1458,6 +1442,13 @@ def add_dependency_edge(self, dependency_spec, deptype):
             dependency_spec (Spec): spec of the dependency
             deptype (str or tuple): dependency types
         """
+        # Check if we need to update edges that are already present
+        selected = self._dependencies.select(child=dependency_spec.name)
+        for edge in selected:
+            if id(dependency_spec) == id(edge.spec):
+                edge.add_type(deptype)
+                return
```

Comment:
@tgamblin I still need to update this part, to do the symmetric operation on the parent It seems though we have no test that detected this.